### PR TITLE
drivers/periph: add interface for QSPI (API only)

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

This adds the API for the QSPI peripheral found on many chips.
The border between the external memory chip and the internal QSPI chip is a bit mushy at times when the QSPI peripheral implements part of the JEDEC protocol.

### Testing procedure

Read the API and verify that it makes sense.
There is also a minimal test application that can be used to verify the implementation.

### CODE REVIEW

File: drivers/include/periph/qspi.h [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+ * @param[in] bus       QSPI device to release
+ */
+void qspi_release(qspi_t bus);
+
+/**
+ * @brief   Execute a command with response data e.g Read Status, Read JEDEC
+ *
+ * @param[in]  bus      QSPI device
+ * @param[in]  command  The JEDEC command ID
+ * @param[out] response Command response data, may be NULL
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:182]: What's the expected behaviour when `response` is `NULL` and `len` is not `0`? Or is this illegal?

File: drivers/include/periph/qspi.h [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+ * @param[in] len       Size of the command response buffer
+ */
+void qspi_cmd_write(qspi_t bus, uint8_t command, const void *data, size_t len);
+
+/**
+ * @brief   Read data from external memory.
+ *          Typically it is implemented by quad read command (`0x6B`).
+ *
+ * @param[in]  bus      QSPI device
+ * @param[in]  address  Address to read from
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:202]: The nrf52 expects the address to be word-aligned. Is that the same with the same54? Should that be enforced at API-level?

File: drivers/include/periph/qspi.h [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+ * @param[out] data     Buffer to store the data
+ * @param[in]  len      Nmber of byte to read
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:204]: And the nrf52 also expects a multiple of 4 for the `len`.

Comment (bergzand) [inline drivers/include/periph/qspi.h:204]:
```suggestion
 * @param[in]  len      Number of byte to read
```

File: drivers/include/periph/qspi.h [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+/**
+ * @brief   Write data to external memory.
+ *          Can only write 1 -> 0, so target region should erased first.
+ *          Typically it uses quad write command (`0x32`).
+ *
+ * @param[in] bus       QSPI device
+ * @param[in] address   Address to write to
+ * @param[in] data      Buffer to write
+ * @param[in] len       Nmber of byte to write
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:216]:
```suggestion
 * @param[in] len       Number of byte to write
```

File: drivers/include/periph/qspi.h [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+/**
+ * @brief   Configure the QSPI peripheral settings
+ *
+ * @param[in] bus       QSPI device to configure
+ * @param[in] mode      QSPI mode @see qspi_mode_t
+ * @param[in] flags     QSPI Configuration Options
+ * @param[in] clk_hz    QSPI frequency in Hz
+ */
+void qspi_configure(qspi_t bus, qspi_mode_t mode, uint32_t flags, uint32_t clk_hz);
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:153]: How should a peripheral implementation handle invalid configuration settings? Throw an assertion failure or return an error code. For example, the nrf52 only supports `qspi_mode_t` 0 and 3. It also only supports clock speeds between 2 MHz and 32 MHz.

File: drivers/include/periph/qspi.h [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+ * @param[in] data      Buffer to write
+ * @param[in] len       Nmber of byte to write
+ */
+void qspi_write(qspi_t bus, uint32_t addr, const void *data, size_t len);
+
+/**
+ * @brief   Erase external memory by address
+ *
+ * @param[in] bus       QSPI device
+ * @param[in] address   The address of the block that will be reased
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:224]:
```suggestion
 * @param[in] address   The address of the block that will be erased
```

**Review by bergzand — COMMENTED** [review]
Initial set of questions and typo fixes

**Comment by bergzand** [comment]
I have a somewhat working (very minimal) implementation of the nrf52 peripheral.
What I noticed so far in terms of restrictions:
1. Data read from and written to the flash must reside in RAM (EasyDMA limitation).
2. Data pointers must be word aligned (another EasyDMA limitation here).
3. Lengths must be multiple of word sizes (Is this a JEDEC spec thing?).
4. Flash addresses must be word aligned (probably a JEDEC spec thing).

I assume that 3 and 4 are per JEDEC spec and shouldn't cause issues. Throwing an assert on this in the implementation is probably fine.

1 can be solved within the peripheral driver. There is already a similar check in the plain SPI implementation to copy data to a temporary buffer before transmitting. Something similar can be done here.
As for the alignment issue (number 2), I'm not sure how to handle this, the solution for issue 1 can be used. Zephyr seems to solve this by piecewise copying the data with an optional `memmove` in to account to copy at least the bulk of the data in a single transfer.

**Comment by bergzand** [comment]
[Here](https://github.com/bergzand/RIOT/tree/pr/nrf52/qspi) is my branch btw. I will PR it as soon as it is cleaned up and I'm somewhat confident in it.

**Comment by benpicco** [comment]
The QSPI peripheral on the samd5x does unaligned reads / writes just fine.
I think for a 4 byte intermediate buffer, we can do the full-word stuffing inside the driver if needed. That's what the [Adafruit_SPIFlash](https://github.com/adafruit/Adafruit_SPIFlash/blob/master/src/qspi/Adafruit_FlashTransport_QSPI_NRF.cpp#L159) does for nRF and that's what I did for writing the [config area](https://github.com/RIOT-OS/RIOT/pull/14512/files#diff-22252051f7ceac9462d1f1f181219f1a376a88a31001bc7079ff804f188ba190R204) on sam0 which also has this requirement.

I think we should make the qspi_read() and qspi_write() functions return the number of bytes written, so if a peripheral only supports a maximal transfer size, the MTD layer will take care of the chunking.

File: drivers/include/periph/qspi.h (hunk 2 above)
Comment (benpicco) [reply drivers/include/periph/qspi.h:202]: That seems to be a limitation of the nrf52.
I'd say the driver can best deal with that :wink:

File: drivers/include/periph/qspi.h (hunk 1 above)
Comment (benpicco) [reply drivers/include/periph/qspi.h:182]: If `response` is NULL, `len` would be ignored.

File: drivers/include/periph/qspi.h (hunk 5 above)
Comment (benpicco) [reply drivers/include/periph/qspi.h:153]: The question is: Can those errors be handled at run-time?
If the static configuration is wrong, I think crashing with a failed assertion is fine as that would only happen during development.

**Comment by bergzand** [comment]
Another thing I would like to see in this API is support for [XIP](https://en.wikipedia.org/wiki/Execute_in_place) offered by many of the QSPI peripherals. This would allow us to run firmware from the flash chip directly

File: drivers/include/periph/qspi.h [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+ */
+void qspi_release(qspi_t bus);
+
+/**
+ * @brief   Execute a command with response data e.g Read Status, Read JEDEC
+ *
+ * @param[in]  bus      QSPI device
+ * @param[in]  command  The JEDEC command ID
+ * @param[out] response Command response data, may be NULL
+ * @param[in]  len      Size of the command response buffer
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:183]: Do we have a realistic limit for this? The nRF52 has a 18 bit length register for the read and write lengths, or about 256KiB. Of course I can handle the split in the driver, just wondering if it is necessary to handle it in the driver.

**Comment by bergzand** [comment]
> Another thing I would like to see in this API is support for [XIP](https://en.wikipedia.org/wiki/Execute_in_place) offered by many of the QSPI peripherals. This would allow us to run firmware from the flash chip directly

I think for this it is sufficient to have an

```
void qspi_mount_xip(qspi_dev_t dev, void **region_start);
```

Should of course also be acquired and released as I think we should disallow writes while the xip mount is active.

**Comment by benpicco** [comment]
So that would simply return the memory to which the QSPI is mapped?

**Comment by bergzand** [comment]
> So that would simply return a pointer to the memory to which the QSPI is mapped?

That's the idea for now

**Comment by bergzand** [comment]
Maybe it makes sense to add a separate feature flag for XIP support: `periph_qspi_xip`

File: drivers/include/periph/qspi.h (hunk 7 above)
Comment (benpicco) [reply drivers/include/periph/qspi.h:183]: Nah we should just return the number of bytes read.
For commands I don't expect large buffers anyway. (and they have to be in memory *somewhere*)

Comment (bergzand) [reply drivers/include/periph/qspi.h:183]: Of course, makes sense.

File: drivers/include/periph/qspi.h (hunk 1 above)
Comment (benpicco) [reply drivers/include/periph/qspi.h:182]: I can drop the 'may be NULL' if that makes thinks simpler.

Comment (bergzand) [reply drivers/include/periph/qspi.h]: I think that simplifies things, keep reads for reading data and only have the `data == NULL` edge case in the write function

File: drivers/include/periph/qspi.h [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+#define QSPI_FLAG_2BIT          (0x1)   /**< two parallel data lines    */
+#define QSPI_FLAG_4BIT          (0x2)   /**< four parallel data lines   */
+#define QSPI_FLAG_8BIT          (0x3)   /**< eight parallel data lines  */
+
+#define QSPI_FLAG_BIT_MASK      (0x3)   /**< mask to get data lines from flags */
+#define QSPI_FLAG_BIT(flags)    ((flags) & QSPI_FLAG_BIT_MASK)
+
+#define QSPI_FLAG_DDR           (0x4)   /**< use Double Data Rate mode  */
+
+#define QSPI_FLAG_ADDR_32BIT    (0x8)   /**< use 32 bit addressing      */
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:133]: This also needs a flag to signal that the address can also be spread over the data lines. In both the nrf52 and the stm32 peripheral this is a separate configuration setting and behaviour of various memory chips are slightly different when the address is also spread over the available data lines.

```suggestion
#define QSPI_FLAG_ADDR_32BIT    (0x8)   /**< use 32 bit addressing      */
#define QSPI_FLAG_ADDR_MULTILINE  (0x10) /**< Use parallel data lines for the address */
```

File: drivers/include/periph/qspi.h [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+ * @brief   Configure the QSPI peripheral settings
+ *
+ * @pre     The @p bus has been acquired with @ref qspi_acquire
+ *
+ * @param[in] bus       QSPI device to configure
+ * @param[in] mode      QSPI mode @see qspi_mode_t
+ * @param[in] flags     QSPI Configuration Options
+ * @param[in] clk_hz    QSPI frequency in Hz
+ */
+void qspi_configure(qspi_t bus, qspi_mode_t mode, uint32_t flags, uint32_t clk_hz);
```
Comment (bergzand) [inline drivers/include/periph/qspi.h:155]: I have a memory chip here that requires a number of dummy cycles between the address and the data. Should we include that setting in the configure function here? Both the nrf52 and the stm32 peripheral can be configured to emit a few dummy bytes into the read and write calls.

**Comment by bergzand** [comment]
Is there a standardized setting to enable quad I/O on these flash chips? Or should it be enabled by default? I have a winbond W25Q16BV and a Micron n25Q256a here and they seem to have different configuration registers to enable quad SPI operation.

**Comment by benpicco** [comment]
> Is there a standardized setting to enable quad I/O on these flash chips? Or should it be enabled by default? I have a winbond W25Q16BV and a Micron n25Q256a here and they seem to have different configuration registers to enable quad SPI operation.

That's what I was wondering too.
It seems like there are two ways to do QSPI

 - send a Quad Enable command and then use the normal Fast Read command, etc.
 - use Quad Read command instead of Fast Read, no additional commands needed

**Pull request closed without merging**